Documentation/Evolution/DelimiterSyntax.md

```diff
@@ -0,0 +1,165 @@
+# Regular Expression Literal Delimiters
+
+- Authors: Hamish Knight, Michael Ilseman
+
+## Introduction
+
+**TODO**
+
+**TODO: Motivation for regex literals in the first place? Or is that a given?**
+
+**TODO: Overview of regex literals in other languages?**
+
+## Detailed Design
+
+A regular expression literal will be introduced using `re'...'` delimiters, within which the compiler will parse a regular expression (the details of which are outlined in **TODO: Internal syntax link**):
+
+```
+// Matches "<identifier> = <hexadecimal value>", extracting the identifier and hex number
+let regex = re'([[:alpha:]]\w*) = ([0-9A-F]+)'
+```
+
+The use of a two letter prefix allows for easy future extensibility of such literals, by allowing different prefixes to indicate different types of literal. **TODO: examples**
```

Comment:
(super early musings). I think it's interesting to contrast single-letter, double-quotes, and full spelling (e.g. `regex'...'`).

In my mind, we're proposing the following world:

1. Regexes are special enough to be "blessed" in some fashion
2. Regexes are not uniquely special though, other foreign algorithm literals are similarly arguable
3. We want to keep the possibility of supporting library-provided foreign algorithm literals in the (distant?) future

Single-letter namespaces run out fast, and it's not the clearest in the world if `r'...'` means raw or regex, while `re'...'` is very clear. `regex'...'` might be even more clear, but it feels (at least to me) overdone.

I think future extension to an extended Swift regex syntax using `rx'...'` is a nice-to-have and can serve as a clear example of extensibility. But, it's not super compelling on its own.

I don't have a lot of experience using other embedded languages like graphql or sql, so I'm not sure how those would play out. I don't know if there's something similar (i.e. a compact common notation) in the numerics world (@stephentyrone). 

Double quotes seem good on the surface, but these are very much not strings. Backslashes mean something completely different, etc. These are _foreign program literals_ not _data literals_. Raw strings are a way to write run-time regexes for use with NSRegularExpression. But I haven't really thought this argument out a ton either.

Reply:
I wonder about `sh'...'` for shelling out, etc.